gnomad_qc/v4/create_release/update_hgdp_tgp_subset_AF.py

```diff
@@ -0,0 +1,484 @@
+                .when(samples_contaminated.contains(ht.s), True)
+                .when(samples_hard_filtered.contains(ht.s), True)
+                .default(False)
+            )
+        )
+    )
+    num_hard_filtered_old = ht.filter(ht.gnomad_sample_filters.hard_filtered).count()
+    num_hard_filtered_new = ht.filter(
+        ht.gnomad_sample_filters.hard_filtered_updated
+    ).count()
+
+    logger.info("%d sample(s) hard filtered", num_hard_filtered_new)
+    logger.info(
+        "%d more sample(s) hard filtered than last release",
+        num_hard_filtered_new - num_hard_filtered_old,
+    )
+
+    relatedness_ht = hgdp_tgp_related_samples_to_drop.ht()
+    samples_related = relatedness_ht.node.s.collect(_localize=False)
+    ht = ht.annotate(
+        relatedness_inference=ht.relatedness_inference.annotate(
+            related_updated=hl.if_else(samples_related.contains(ht.s), True, False)
+        )
+    )
+
+    logger.info(
+        "%d sample(s) related",
+        ht.aggregate(hl.agg.count_where(ht.relatedness_inference.related_updated)),
+    )
+    logger.info(
+        "%d sample(s) were assigned differently for the relatedness than last release",
+        ht.aggregate(
+            hl.agg.count_where(
+                ht.relatedness_inference.related_updated
+                != ht.relatedness_inference.related
+            )
+        ),
+    )
+
+    pop_outliers_ht = hgdp_tgp_pop_outliers.ht()
+    samples_pop_outliers = pop_outliers_ht.s.collect(_localize=False)
+    ht = ht.annotate(
+        hgdp_tgp_meta=ht.hgdp_tgp_meta.annotate(
+            subcontinental_pca=ht.hgdp_tgp_meta.subcontinental_pca.annotate(
+                outlier_updated=hl.if_else(
+                    samples_pop_outliers.contains(ht.s), True, False
+                )
+            )
+        )
+    )
+
+    logger.info(
+        "%d sample(s) are pop outliers",
+        ht.aggregate(
+            hl.agg.count_where(ht.hgdp_tgp_meta.subcontinental_pca.outlier_updated)
+        ),
+    )
+    logger.info(
+        "%d sample(s) were assigned differently as subcontinental outlier than last"
+        " release",
+        ht.aggregate(
+            hl.agg.count_where(
+                ht.hgdp_tgp_meta.subcontinental_pca.outlier_updated
+                != ht.hgdp_tgp_meta.subcontinental_pca.outlier
+            )
+        ),
+    )
+
+    populations_ht = hgdp_tgp_populations_updated.ht()
+    ht = ht.annotate(
+        hgdp_tgp_meta=ht.hgdp_tgp_meta.annotate(
+            population_updated=populations_ht[ht.s].population,
+            latitude_updated=populations_ht[ht.s].latitude,
+            longitude_updated=populations_ht[ht.s].longitude,
+        ),
+    )
+    logger.info(
+        "%d samples have different population labels than last release",
+        ht.aggregate(
+            hl.agg.count_where(
+                ht.hgdp_tgp_meta.population_updated != ht.hgdp_tgp_meta.population
+            )
+        ),
+    )
+    return ht
```

Comment:
> I made some changes that I think help with seeing exactly what is changing and organization.
> 
> * I like to load all of the tables at the top of the function.
> * While we are at it, I think we should just update the contamination rate too
> * There are a lot of collect and counts that don't need to be done because it is just causing things to be recomputed several times when they don't need to be.
> * I thought it would be good to add an annotation indicating changes have been made, so it's obvious which have changed for a given sample and can also be used later in the code.
> 
> Questions that I didn't address with my code changes:
> 
> * We will want to update relatedness_inference.related_samples as well, so either reach out to Mary to get info, or add a TODO: so we can address this for the subset release that will come after the v4 MVP
- **Qin: for this info, it seems she didn't checkpoint the relatedness_ht like last time, and they didn't update the meta in their final MTs, they only returned related samples IDs, do we have to rerun it and checkpoint it?**
> * More things that we should either change or add a TODO for:
>   
>   * hgdp_tgp_meta.global_pca_scores
>   * hgdp_tgp_meta.subcontinental_pca.pca_scores
>   * hgdp_tgp_meta.subcontinental_pca.pca_scores_outliers_removed
- **Qin:  I could get the paths from her notebook, are we using the scores with or without outliers? Here: https://github.com/atgu/hgdp_tgp/blob/master/tutorials/nb2.ipynb**

> * I added a TODO for this, we need to decide on what to do about the gnomad_sample_filters.release_related filter. One option we could present is to get another related annotation indicating if a sample is related to another gnomAD release sample that is not in the HGDP + 1KG subset.
- **Qin: I thought we have to rerun cuKing on the genomes, because it seems pc_relate was overestimating the relatedness for some small populations, Papuan were related to native Americans**
> * The  `_update_annotations` function I added could be modified a little and put into gnomad_methods, it might be useful for others if they need to update annotations that are highly structured like this.
> * I gave up on thinking about names, so you will need to change names in the `_update_annotations` and add docstrings.
> * I think the logger at the end could probably be improved, but it's not really a priority.
- **Qin: I will try to work on this, I wasn't sure if we need to have a separate column or just update the old one, so we could keep track on what's changed"

Reply:
> Qin: for this info, it seems she didn't checkpoint the relatedness_ht like last time, and they didn't update the meta in their final MTs, they only returned related samples IDs, do we have to rerun it and checkpoint it?

We will need to for the subset release yeah, but you don't need to add it to this now, it can just be a TODO that we come back to after the MVP. However, it might be good to write out the file and save it now just to confirm we get the same final output.

> Qin: I could get the paths from her notebook, are we using the scores with or without outliers? Here: https://github.com/atgu/hgdp_tgp/blob/master/tutorials/nb2.ipynb

Same thought here, just make sure we have the files, but you can add a TODO for post v4 MVP. We add all scores to the metadata HT.

> Qin: I thought we have to rerun cuKing on the genomes, because it seems pc_relate was overestimating the relatedness for some small populations, Papuans were related to native Americans so they will not have frequency in the whole release. In general, I think this one might be problematic.

I don't think we will run cuKING, we can discuss this more later, it's a topic for the whole group. Actually @ch-kr do you think this is a project meeting topic, or just DSP/ATGU meeting topic?

> Qin: I will try to work on this, I wasn't sure if we need to have a separate column or just update the old one, so we could keep track on what's changed

Yeah, makes sense, I think this is a clearer way to handle it

Reply:
> Qin: I thought we have to rerun cuKing on the genomes, because it seems pc_relate was overestimating the relatedness for some small populations, Papuans were related to native Americans so they will not have frequency in the whole release. In general, I think this one might be problematic.

> Julia: I don't think we will run cuKING, we can discuss this more later, it's a topic for the whole group. Actually @ch-kr do you think this is a project meeting topic, or just DSP/ATGU meeting topic?

To make sure I'm following:

- Alicia's team generated new relatedness results for the HGDP/1KG subset, so we want to update the `relatedness_inference` struct
- Discussing `relatedness_inference.related_samples` led to Qin's question ("I thought we have to rerun cuKing")

I also do not think we should rerun cuKING, but if we need to discuss whether to do so, the project meeting seems like a more appropriate venue

Reply:
@ch-kr yes, they reran pc_relate with a different pipeline, and they get a somewhat different set of related samples within HGDP/1KG subset compared to v3 subset, we are going to use the new one. 
What we need to discuss is that how to define how HGDP/1KG are related to other gnomAD.samples, it was this `gnomad_sample_filters.release_related` that we are concerned about.

Reply:
thanks for clarifying! sounds like a project meeting topic.

I'm inclined to take whichever path means we don't have to redo any QC on v3

Reply:
Sounds good to me! @ch-kr the in silico predictors shouldn't take a lot of time I hope.